agent/agent.go

```diff
@@ -254,17 +275,44 @@ func (a *Agent) Start() error {
 		return fmt.Errorf("Failed to setup node ID: %v", err)
 	}
 
+	// create a notif channel to trigger state sychronizations
+	// when a consul server was added to the cluster.
+	serverUpCh := make(chan struct{}, 1)
+
+	// create a notif channel to trigger state synchronizations
+	// when the state has changed.
+	triggerCh := make(chan struct{}, 1)
+
 	// create the local state
-	a.state = NewLocalState(c, a.logger, a.tokens)
+	a.State = local.NewState(LocalConfig(c), a.logger, a.tokens, triggerCh)
+
+	// create the state synchronization manager which performs
+	// regular and on-demand state synchronizations (anti-entropy).
+	a.sync = &ae.StateSyncer{
+		State:      a.State,
+		Interval:   c.AEInterval,
+		ShutdownCh: a.shutdownCh,
+		ServerUpCh: serverUpCh,
+		TriggerCh:  triggerCh,
+		Logger:     a.logger,
+	}
 
 	// create the config for the rpc server/client
 	consulCfg, err := a.consulConfig()
 	if err != nil {
 		return err
 	}
 
-	// link consul client/server with the state
-	consulCfg.ServerUp = a.state.ConsulServerUp
+	// ServerUp is used to inform that a new consul server is now
+	// up. This can be used to speed up the sync process if we are blocking
+	// waiting to discover a consul server
+	// todo(fs): IMO, the non-blocking nature of this call should be hidden in the syncer
```

Comment:
That's not a bad idea to just call `local.NotifyNewServer()` instead of plumbing a channel here.

Reply:
The problem is that you need the non-blocking nature in two places: in the local state and the AE code. If we want to hide this we have two options: callback or move the nonblocking feature into a separate struct. The latter feels like gold-plating whereas the former works against the decoupling. In essence, we want a channel which is not blocking on write.

Reply:
I've added an `ae.Trigger` which hides this implementation detail and wire them up with callbacks between the state, syncer and the consul rpc server.